pkg/ipcache/metrics.go

```diff
@@ -0,0 +1,31 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipcache
+
+import (
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/metrics"
+)
+
+const (
+	metricsSubsystem = "ipcache"
+)
+
+func metricsKvStoreIncrementEvent(eventType kvstore.KeyValueEvent) {
```

Comment:
This is still WIP, but I would like to have a `metrics` tool per each package. In this case is the same, but  as long as we grow in metrics, the logic about the metrics are in the same package/file, and we keep the logic in the same place, so it's easy to track where the metrics are called. What do you think? 

The thing is, if we made a lot of methods in metrics pkg the number of Exported helpers will be used across all packages, and the responsibility will be in the wrong place, the implemenation of the metrics should be per pkg. IMHO.

Reply:
I see. I was thinking from a code duplication point of view. Every duplication of a function is another place which might introduce a bug, or also needs to be updated if there is a change later on. I note that already in this PR rather than adding the new file and creating another copy of the function, one of the call points (pkg/node) already open-codes the same logic.

Most metrics bits should be a simple matter of declaring a map[string]string inside the caller package to specify the labels, then doing something like `metrics.MyMetric.With(labelsFromMyPackage).Inc()`. I think this achieves the split you're after, but doesn't work quite as well in a case like this where the metric label depends on a local variable.

IMO If the same metric is used from multiple packages, code for it should live in `pkg/metrics`; if it's only used from one package, then most of the code should be in `pkg/foo/metrics.go`.